src/dune_rules/compilation_context.ml

```diff
@@ -175,6 +179,38 @@ let create ~super_context ~scope ~expander ~obj_dir ~modules ~flags
     ; stdlib
     }
   in
+  let melange_js_includes =
+    match melange with
+    | None -> None
+    | Some (melange_stanza_dir, target_dir) ->
+      let open Resolve.Memo.O in
+      Some
+        (Command.Args.memo
+           (Resolve.Memo.args
+              (let+ libs = requires_compile in
+               let project = Scope.project scope in
+               let deps_of_lib (lib : Lib.t) ~groups =
+                 let lib = Lib.Local.of_lib_exn lib in
+                 let info = Lib.Local.info lib in
+                 let lib_dir = Lib_info.src_dir info in
+                 let dst_dir =
+                   Melange.lib_output_dir ~melange_stanza_dir ~lib_dir
+                     ~target:target_dir
+                 in
+                 List.map groups ~f:(fun g ->
+                     let dir = Path.build dst_dir in
+                     Lib_file_deps.Group.to_predicate g
+                     |> File_selector.create ~dir |> Dep.file_selector)
+                 |> Dep.Set.of_list
+               in
+               let deps libs ~groups =
+                 Dep.Set.union_map libs ~f:(deps_of_lib ~groups)
+               in
+               Command.Args.S
+                 [ Lib_flags.L.include_flags ~project libs Melange
+                 ; Hidden_deps (deps libs ~groups:[ Melange Js ])
```

Comment:
is this expected? does melange actually read other .js files when it's outputting .js?

Reply:
It shouldn't be necessary.

Reply:
I am confused. Without these deps on the js files, how would these `.js` targets get included in the build?

To give an example, if I have `app/a.ml` that depends on some other library module `lib/b.ml`, there will be a `require('../lib/b.js')` in the resulting `app/a.js` file. Without these hidden dependencies, or if we depended only on `.cmj` user would get `Error: Cannot find module '../lib/b.js'` when trying to run Node or some bundler like Webpack.

Reply:
In other words, Melange does not depend on the .js files, but the dependency is there.

Reply:
I would assume they would just be there if they're sub-targets of the `emit` stanza.

Reply:
> I would assume they would just be there if they're sub-targets of the emit stanza.

I am not sure I understand what this means. Sub-targets of what other target exactly? The only "entry" target in the emit stanza are the libraries listed.

To clarify, this is how things work at the moment: in the `emit` stanza, user includes some libraries as direct deps in the `libraries` field. We iterate over the sources of these libs here, to add their `.js` as targets:

https://github.com/ocaml/dune/blob/295dd775a5f9aa251807e579e51db265a83ea37a/src/dune_rules/melange_rules.ml#L74-L77

Now, these direct dependencies can depend on other libraries. At some point we have to include their `.js` artifacts as targets of the build too. This is what the code in line 211 above is doing.

I guess we could iterate upfront over all transient dependencies and include their `.js` targets in the build, is that what you have in mind? I would argue the current approach is more concise and less prone to errors as they are added organically at the time where the `-I` flag is used, in a similar way to what is done for other artifacts like `cmo` or `cmj`.

Reply:
I meant that every individual `.js` target only needs to depend on its `.cmj` (and dependencies of that `.cmj`, transitively).

You don't need to depend on `A.js` to emit `B.js`, because the `emit` stanza guarantees the `A.js` emission. Is this not the correct mental model?

Reply:
Just to clarify, the issue isn't that we should .js dependencies somewhere. We clearly do, because as you've correctly stated they will not be copied to the target directory without this being stated somewhere. But there are serious problems with adding unnecessary dependencies to targets:

* The dune cache is harmed. The key the dune cache uses includes all the dependencies. When this key is computed using unnecessary dependencies, the opportunities for hitting the cache are diminished.
* Dune's schedules a bad build plan. Since dune builds targets in dependency order, unnecessary dependencies enforce serialization constraints and reduce possibilities of running actions concurrently.

So however elegant it may seem, it is not desirable.

> I guess we could iterate upfront over all transient dependencies and include their .js targets in the build, is that what you have in mind?

This approach is the best IMO.

Note that an even better approach is possible if we introduce another constraint. If the user provides an `entry_module` field, dune can setup rules by climbing the transitive closure of this module and only setup the minimal number of .js rules.

Reply:
I understand the issues that you mention. But this seems to be a common problem of any artifacts generated as part of the library build? To showcase what I mean, I created a cram test with a byte and native build example: https://github.com/ocaml/dune/pull/6364.

> Note that an even better approach is possible if we introduce another constraint. If the user provides an `entry_module` field,

This would fix the problem for the entry library only, but not for libraries that the first one depends on, which seems like the main issue?

Tbh I am not sure I am understanding yet what the issue is, nor how it is specific to Melange.

Reply:
> I understand the issues that you mention. But this seems to be a common problem of any artifacts generated as part of the library build? To showcase what I mean, I created a cram test with a byte and native build example: https://github.com/ocaml/dune/pull/6364.

It's not the same issue. Your test assumes that `foo__Y` is unnecessary because it's never referenced in the executable code. But that's now how dune compiles against libraries. When we build against libraries in OCaml, we build against their archive files - in this case `foo.cma`. To produce this archive file, we indeed need this unused module `foo__Y`.

The issue I'm pointing out is totally specific to melange. There are no analogous rules needed to build anything in OCaml.

Reply:
I see, I think I get it now, thanks.

So, the rough plan would be to:
- add new field `entry_module` to `melange.emit` stanza
- update `libraries` field to only allow 1 library and be renamed `library` otherwise I believe it'd lead to confusing behavior regarding `entry_module`
- when `melange.emit` is processed, if the fields are `(library foo)` and `(entry_module bar)` we would start walking from `Foo.bar`, get all dependency modules, processing until no more module deps are available (what was referred as the "transitive closure of this module"?). And add a `.js` rule for each of the resulting modules.

This triggers some questions, like:
- what about performance? Can we use Dune cache or some other mechanisms to store the results of this dependency walk? I suspect it might get expensive for some projects.
- For `cmj` we are using a similar approach than the one being discussed for `js` artifacts in this thread. I understand that `cmj` rule generation should not be modified to follow the new, more performant approach? i.e. we will still generate `cmj` rules even if they are not needed?
- would this change affect 3rd party installation of libraries in any way?

Reply:
Actually, if replacing `libraries` with `library` makes sense, I am wondering if it'd be easier to to just replace `melange.emit` with a new field `melange_emit` added to `library` stanza, with the following schema:

```
(library
  (name foo)
  (melange_emit (target output) (entry_module bar) (spec es6)))
```

This would produce artifacts in `_build/default/<path_to_foo_lib>/output/`, and would reduce a lot of the complexity of handling edge cases due to `melange.emit` and its libraries being in separate folders, as it won't be possible anymore.

Another alternative with roughly the same upsides would be to just remove the `libraries` field:

```scheme
(melange.emit (target output) (entry_module bar) (spec es6))
```

The `melange.emit` stanza then would target the module `bar.ml` placed in the same folder where the stanza exists.


Wdyt?

Reply:
what's the entry module for a library?

Reply:
It's the module from which we will start collecting all dependencies, what @rgrinberg mentioned above:

> Note that an even better approach is possible if we introduce another constraint. If the user provides an `entry_module` field, dune can setup rules by climbing the transitive closure of this module and only setup the minimal number of .js rules.

Reply:
The way I see it, it would be analog to Webpack [entry](https://webpack.js.org/concepts/entry-points/#single-entry-shorthand-syntax).

Reply:
That's a good idea for _executables_. Libraries might ship unrelated stuff that only executables depend on.

Reply:
Right. Effectively, my latest suggestion of `melange.emit` is a kind of executable but for Melange:

> Another alternative with roughly the same upsides would be to just remove the `libraries` field:

```scheme
(melange.emit (target output) (entry_module bar) (spec es6))
```

> The `melange.emit` stanza then would target the module `bar.ml` placed in the same folder where the stanza exists.

Reply:
how do you find libraries to depend on in that case?

Reply:
Hm, right, I got confused. I guess the stanza would still need `libraries`, but the semantics of the field change (imo) quite a bit, from:

- the libraries we want to start producing .js files from

to:

- the libraries that `entry_module` depends upon

I guess this is what Rudi had in mind originally, I will start working then on adding `entry_module` then and "lifting" the js rules generation to the `melange.emit` stanza.

Reply:
I think this is a little weird. In a Next.js application, what's the entry module?

I think we'll have to generate all the `.js` everytime.

Reply:
I guess we can make `entry_module` optional.

- If it's not included, then Dune defines the "starting set" of js rules from all modules that can be found inside the libraries in the `libraries` field (Next.js mode)
- If it's included, Dune would define the "starting set" of js rules from the module itself (Webpack mode)

Reply:
I understand we are trying to solve 2 problems at once in this thread, their outcome is kind of related (wasteful js rules being created) but they are not exactly the same issue:

- problem 1: applies to **direct dependencies only**. Using `libraries` as "entry point" for .js generation means dune might create rules for many `.js` artifacts when user can potentially need just one.
- problem 2: applies to **transitive libraries only**. When some compilation unit (either an entry_module or a library) depends on another library, dune currently generates .js rules for all the modules found in this dependency lib, even if only a bunch of them are used by the original module.

I can see how `entry_module` field fixes 1 as it narrows down the entry points of the dep graph to a single module.

As for 2, I am a bit more confused as to what the fix would be, but I understand we would walk the depgraph from the entry points (either entry_module or libraries) upfront, and use the resulting set of modules to create all the .js rules?

Reply:
> As for 2, I am a bit more confused as to what the fix would be, but I understand we would walk the depgraph from the entry points (either entry_module or libraries) upfront, and use the resulting set of modules to create all the .js rules?

It's not something we need to worry about right now, but the solution is for dune not to over-approximate the dependency graph at the level of libraries. Dune already has the fine grained dependency information at the module level, there's no reason why it has to use the approximation. It would require a little more work on our end in dune. I'd like to address it once we integrate with the codept tool.

Reply:
I see. So we can defer fixing 2 then.

Re: 1, I intend to implement the following solution, as part of this PR.

## `entries` field

Add a new `entries` field to the `melange.emit` stanza that can take a list of values. The list can contain two types of values:

a) **Modules as entries**: `(entries foo bar)` will take modules `foo.ml` and `bar.ml` as entries to generate js rules. It can also be written `(entries (module foo) (module bar))`

b) **Libraries as entries**: `(entries (library x) (library y))` will take all modules in libraries `x` and `y` as entries to generate js rules (the Next.js case that @anmonteiro mentioned above).

They can be mixed together, e.g. `(entries (module foo) (library y))`

That way, we can "free" the `libraries` field to only mean the dependencies to be included in `-I` params, no ambiguities. 

My only hesitation is if case b) above should be `(folder x)` instead 🤔 so that the concept of library remains fully detached from the concept of "entry".

Wdyt?

Reply:
I think you should keep things simple for now and just implement whatever you actually need for your existing codebase. So either a) or b) but not both.

> My only hesitation is if case b) above should be (folder x) instead 🤔 so that the concept of library remains fully detached from the concept of "entry".

I don't think I understand the suggestion. What is `x` in this case?

Reply:
`x` would be the folder that we want to use to gather all entries from. So dune would check all modules in the folder and add js rules for them.

In any case, I followed your suggestion and ended up not adding support for folders (b case) for now, just for explicit modules, see https://github.com/ocaml/dune/pull/6349/commits/648892205b7af6f1c9d4a1ffa4dedee61f03aed4.